shell/platform/android/io/flutter/embedding/engine/android/FlutterSurfaceView.java

```diff
@@ -0,0 +1,159 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.android;
+
+import android.content.Context;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.AttributeSet;
+import android.util.Log;
+import android.view.SurfaceHolder;
+import android.view.SurfaceView;
+
+import java.nio.ByteBuffer;
+
+import io.flutter.embedding.engine.renderer.FlutterRenderer;
+
+/**
+ * Paints a Flutter UI on a {@link android.view.Surface}.
+ *
+ * Consider using a {@link FlutterView}, which utilizes a {@code FlutterSurfaceView}
+ * internally.
+ */
+public class FlutterSurfaceView extends SurfaceView {
+  private static final String TAG = "FlutterSurfaceView";
+
+  private boolean isSurfaceAvailableForRendering = false;
+  private boolean isAttachedToFlutterRenderer = false;
+  @Nullable
+  private FlutterRenderer flutterRenderer;
+
+  // Connects the {@code Surface} beneath this {@code SurfaceView} with Flutter's native code.
+  // Callbacks are received by this Object and then those messages are forwarded to our
+  // FlutterRenderer, and then on to the JNI bridge over to native Flutter code.
+  private final SurfaceHolder.Callback surfaceCallback = new SurfaceHolder.Callback() {
+    @Override
+    public void surfaceCreated(SurfaceHolder holder) {
+      Log.d(TAG, "SurfaceHolder.Callback.surfaceCreated()");
+      isSurfaceAvailableForRendering = true;
+
+      if (isAttachedToFlutterRenderer) {
+        Log.d(TAG, "Already attached to renderer. Notifying of surface creation.");
+        connectSurfaceToRenderer();
+      }
+    }
+
+    @Override
+    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
+      if (isAttachedToFlutterRenderer) {
+        changeSurfaceSize(width, height);
+      }
```

Comment:
Is it an error state worth logging about if we're not attached to the renderer at this point? Or is it expected that some calls may come through when this isn't attached?

Applies to the `TextureView` listener too.

Reply:
the truth is, I don't know.  It shouldn't ever happen, but if it ever does then I'd need to see it in action to understand what's going on.  Mostly I think we'd like to avoid crashing here.  Maybe there was a slight timing mismatch that caused this method be invoked after the native side was detached and cleaned up.  That's not great but its also not worth crashing over.

Reply:
Actually I took a closer look at this and I think we'll expect this situation to happen a lot.  The surface change callback will be invoked whenever the Android UI changes layout, but it's very possible that the developer has not yet attached any FlutterEngine/Renderer to this View.  So then we'd be spamming logs for a situation that's totally expected.  Does that make sense?

Reply:
Yeah, in that case it makes sense not to log.